File: solutions/ruby/meetup/2/meetup.rb
# frozen_string_literal: true

require 'date'

# class Meetup
class Meetup
  def initialize(month, year) = (@month = month
                                 @year = year)

  def day(weekday, nth)
    d = { first: 1, second: 8, third: 15, fourth: 22, last: -7, teenth: 13 }[nth]
    (Date.new(@year, @month, d)..).find { |date| date.public_send("#{weekday}?") }
  end
end

File: solutions/lua/meetup/2/meetup.lua
local Weekday = { "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday" }
return function(dt)
	local order = { first = 1, second = 2, third = 3, fourth = 4 }
	local days_set = {}
	local days = {}
	for day = 1, 31 do
		local tt = os.date("*t", os.time({ year = dt.year, month = dt.month, day = day }))
		if Weekday[tt.wday] == dt.day and not days_set[tt.day] then
			days_set[tt.day] = true
			days[#days + 1] = tt.day
		end
	end
	if dt.week == "last" then return days[#days] end
	if dt.week == "teenth" then
		for i = 13, 19 do
			if days_set[i] then return i end
		end
	end
	return days[order[dt.week]]
end
